lib/gpio.cc

```diff
@@ -244,25 +244,54 @@ static int ReadFileToBuffer(char *buffer, size_t size, const char *filename) {
   return r;
 }
 
-static RaspberryPiModel DetermineRaspberryModel() {
+/*
+ * Try to read the revision from /proc/cpuinfo. In case of any errors, or if
+ * /proc/cpuinfo simply contains zero as the revision, this function returns
+ * zero. This is ok because zero was never used as a real revision code.
+ */
+static uint32_t ReadRevisionFromProcCpuinfo() {
   char buffer[4096];
   if (ReadFileToBuffer(buffer, sizeof(buffer), "/proc/cpuinfo") < 0) {
     fprintf(stderr, "Reading cpuinfo: Could not determine Pi model\n");
-    return PI_MODEL_3;  // safe guess fallback.
+    return 0;
   }
   static const char RevisionTag[] = "Revision";
   const char *revision_key;
   if ((revision_key = strstr(buffer, RevisionTag)) == NULL) {
     fprintf(stderr, "non-existent Revision: Could not determine Pi model\n");
-    return PI_MODEL_3;
+    return 0;
   }
   unsigned int pi_revision;
   if (sscanf(index(revision_key, ':') + 1, "%x", &pi_revision) != 1) {
     fprintf(stderr, "Unknown Revision: Could not determine Pi model\n");
-    return PI_MODEL_3;
+    return 0;
+  }
+  return pi_revision;
+}
+
+static uint32_t read_be32(const uint8_t *p) {
+  return p[0] << 24 | p[1] << 16 | p[2] << 8 | p[3];
+}
+
+static uint32_t ReadRevisionFromDeviceTree() {
+  char buffer[4];
+  if (ReadFileToBuffer(buffer, sizeof(buffer), "/proc/device-tree/system/linux,revision") < 0) {
```

Comment:
Note, `ReadFileToBuffer()` is meant to read a text file, so it reads one byte less than in the buffer and appends a nul-termination. So for the 4 byte linux revision to work , you need to give it a 5 byte buffer.

Also, you're checking for negative return value here, i.e. if it _wasn't_ able to read. Wouldn't you want to check for `== 4` here ?

Reply:
Good points, I'll review the available I/O functions and test again